DashMap with custom hasher can't be (De)serialized

* dashmap v4.0.1
* serde v1.0.118
* `rustc 1.48.0 (7eac88abb 2020-11-16)`

# Working example:
```rust
use dashmap::DashMap;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct BugExample {
    inner: DashMap<String, String>,
}
```

# Fails to compile:
```rust
use dashmap::DashMap;
use serde::{Deserialize, Serialize};
use std::hash::BuildHasherDefault;
use twox_hash::XxHash64;

#[derive(Serialize, Deserialize)]
pub struct BugExample {
    inner: DashMap<String, String, BuildHasherDefault<XxHash64>>,
}
```

```
error[E0277]: the trait bound `DashMap<std::string::String, std::string::String, BuildHasherDefault<XxHash64>>: Serialize` is not satisfied       
 --> src\main.rs:8:5
  |
8 |     inner: DashMap<String, String, BuildHasherDefault<XxHash64>>,
  |     ^^^^^ the trait `Serialize` is not implemented for `DashMap<std::string::String, std::string::String, BuildHasherDefault<XxHash64>>`      
  |
  = help: the following implementations were found:
            <DashMap<K, V> as Serialize>
  = note: required by `_::_serde::ser::SerializeStruct::serialize_field`

error[E0277]: the trait bound `DashMap<std::string::String, std::string::String, BuildHasherDefault<XxHash64>>: Deserialize<'_>` is not satisfied 
 --> src\main.rs:8:5
  |
8 |     inner: DashMap<String, String, BuildHasherDefault<XxHash64>>,
  |     ^^^^^ the trait `Deserialize<'_>` is not implemented for `DashMap<std::string::String, std::string::String, BuildHasherDefault<XxHash64>>`
  |
  = help: the following implementations were found:
            <DashMap<K, V> as Deserialize<'de>>
  = note: required by `next_element`

error[E0277]: the trait bound `DashMap<std::string::String, std::string::String, BuildHasherDefault<XxHash64>>: Deserialize<'_>` is not satisfied
 --> src\main.rs:8:5
  |
8 |     inner: DashMap<String, String, BuildHasherDefault<XxHash64>>,
  |     ^^^^^ the trait `Deserialize<'_>` is not implemented for `DashMap<std::string::String, std::string::String, BuildHasherDefault<XxHash64>>`
  |
  = help: the following implementations were found:
            <DashMap<K, V> as Deserialize<'de>>
  = note: required by `next_value`
```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -898,9 +898,9 @@ impl<K: Eq + Hash, V, S: BuildHasher + Clone> Extend<(K, V)> for DashMap<K, V, S
     }
 }
 
-impl<K: Eq + Hash, V> FromIterator<(K, V)> for DashMap<K, V, RandomState> {
+impl<K: Eq + Hash, V, S: BuildHasher + Clone + Default> FromIterator<(K, V)> for DashMap<K, V, S> {
     fn from_iter<I: IntoIterator<Item = (K, V)>>(intoiter: I) -> Self {
-        let mut map = DashMap::new();
+        let mut map = DashMap::default();
 
         map.extend(intoiter);
 

--- a/src/serde.rs
+++ b/src/serde.rs
@@ -1,18 +1,19 @@
 use crate::{DashMap, DashSet};
 use core::fmt;
-use core::hash::Hash;
+use core::hash::{Hash, BuildHasher};
 use core::marker::PhantomData;
 use serde::de::{Deserialize, MapAccess, SeqAccess, Visitor};
 use serde::ser::{Serialize, SerializeMap, SerializeSeq, Serializer};
 use serde::Deserializer;
 
-pub struct DashMapVisitor<K, V> {
-    marker: PhantomData<fn() -> DashMap<K, V>>,
+pub struct DashMapVisitor<K, V, S> {
+    marker: PhantomData<fn() -> DashMap<K, V, S>>,
 }
 
-impl<K, V> DashMapVisitor<K, V>
+impl<K, V, S> DashMapVisitor<K, V, S>
 where
     K: Eq + Hash,
+    S: BuildHasher + Clone,
 {
     fn new() -> Self {
         DashMapVisitor {
@@ -21,12 +22,13 @@ where
     }
 }
 
-impl<'de, K, V> Visitor<'de> for DashMapVisitor<K, V>
+impl<'de, K, V, S> Visitor<'de> for DashMapVisitor<K, V, S>
 where
     K: Deserialize<'de> + Eq + Hash,
     V: Deserialize<'de>,
+    S: BuildHasher + Clone + Default,
 {
-    type Value = DashMap<K, V>;
+    type Value = DashMap<K, V, S>;
 
     fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
         formatter.write_str("a DashMap")
@@ -36,7 +38,7 @@ where
     where
         M: MapAccess<'de>,
     {
-        let map = DashMap::with_capacity(access.size_hint().unwrap_or(0));
+        let map = DashMap::with_capacity_and_hasher(access.size_hint().unwrap_or(0), Default::default());
 
         while let Some((key, value)) = access.next_entry()? {
             map.insert(key, value);
@@ -46,23 +48,25 @@ where
     }
 }
 
-impl<'de, K, V> Deserialize<'de> for DashMap<K, V>
+impl<'de, K, V, S> Deserialize<'de> for DashMap<K, V, S>
 where
     K: Deserialize<'de> + Eq + Hash,
     V: Deserialize<'de>,
+    S: BuildHasher + Clone + Default,
 {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: Deserializer<'de>,
     {
-        deserializer.deserialize_map(DashMapVisitor::<K, V>::new())
+        deserializer.deserialize_map(DashMapVisitor::<K, V, S>::new())
     }
 }
 
-impl<K, V> Serialize for DashMap<K, V>
+impl<K, V, H> Serialize for DashMap<K, V, H>
 where
     K: Serialize + Eq + Hash,
     V: Serialize,
+    H: BuildHasher + Clone,
 {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
@@ -78,13 +82,14 @@ where
     }
 }
 
-pub struct DashSetVisitor<K> {
-    marker: PhantomData<fn() -> DashSet<K>>,
+pub struct DashSetVisitor<K, S> {
+    marker: PhantomData<fn() -> DashSet<K, S>>,
 }
 
-impl<K> DashSetVisitor<K>
+impl<K, S> DashSetVisitor<K, S>
 where
     K: Eq + Hash,
+    S: BuildHasher + Clone,
 {
     fn new() -> Self {
         DashSetVisitor {
@@ -93,11 +98,12 @@ where
     }
 }
 
-impl<'de, K> Visitor<'de> for DashSetVisitor<K>
+impl<'de, K, S> Visitor<'de> for DashSetVisitor<K, S>
 where
     K: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
 {
-    type Value = DashSet<K>;
+    type Value = DashSet<K, S>;
 
     fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
         formatter.write_str("a DashSet")
@@ -107,7 +113,7 @@ where
     where
         M: SeqAccess<'de>,
     {
-        let map = DashSet::with_capacity(access.size_hint().unwrap_or(0));
+        let map = DashSet::with_capacity_and_hasher(access.size_hint().unwrap_or(0), Default::default());
 
         while let Some(key) = access.next_element()? {
             map.insert(key);
@@ -117,21 +123,23 @@ where
     }
 }
 
-impl<'de, K> Deserialize<'de> for DashSet<K>
+impl<'de, K, S> Deserialize<'de> for DashSet<K, S>
 where
     K: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
 {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: Deserializer<'de>,
     {
-        deserializer.deserialize_seq(DashSetVisitor::<K>::new())
+        deserializer.deserialize_seq(DashSetVisitor::<K, S>::new())
     }
 }
 
-impl<K> Serialize for DashSet<K>
+impl<K, H> Serialize for DashSet<K, H>
 where
     K: Serialize + Eq + Hash,
+    H: BuildHasher + Clone,
 {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where

--- a/src/set.rs
+++ b/src/set.rs
@@ -403,9 +403,9 @@ impl<K: Eq + Hash, S: BuildHasher + Clone> Extend<K> for DashSet<K, S> {
     }
 }
 
-impl<K: Eq + Hash> FromIterator<K> for DashSet<K, RandomState> {
+impl<K: Eq + Hash, S: BuildHasher + Clone + Default> FromIterator<K> for DashSet<K, S> {
     fn from_iter<I: IntoIterator<Item = K>>(iter: I) -> Self {
-        let mut set = DashSet::new();
+        let mut set = DashSet::default();
 
         set.extend(iter);
 

